kubernetes-go-guestbook/components/serviceDeployment.go

```diff
@@ -0,0 +1,137 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	appsv1 "github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes/apps/v1"
+	corev1 "github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes/core/v1"
+	metav1 "github.com/pulumi/pulumi-kubernetes/sdk/go/kubernetes/meta/v1"
+	"github.com/pulumi/pulumi/sdk/go/pulumi"
+)
+
+type ServiceDeployment struct {
+	pulumi.ResourceState
+
+	FrontendIP pulumi.StringPtrOutput
+	Deployment *appsv1.Deployment
+	Service    *corev1.Service
+}
+
+type ServiceDeploymentArgs struct {
+	Env         corev1.EnvVarArray
+	Image       pulumi.StringInput
+	Labels      pulumi.StringMapInput
+	Ports       pulumi.IntArrayInput
+	Replicas    pulumi.IntInput
+	Resources   corev1.ResourceRequirementsArgs
+	ServiceType pulumi.StringPtrInput
+}
+
+func NewServiceDeployment(
+	ctx *pulumi.Context, name string, args *ServiceDeploymentArgs, opts ...pulumi.ResourceOption,
+) (*ServiceDeployment, error) {
+	serviceDeployment := &ServiceDeployment{}
+	err := ctx.RegisterComponentResource(
+		"kubernetes-go-guestbook:component:ServiceDeployment",
+		name, serviceDeployment, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	portsOutput := args.Ports.ToIntArrayOutput()
+
+	deploymentPorts := portsOutput.ApplyT(func(ports []int) []corev1.ContainerPort {
+		var result []corev1.ContainerPort
+		for _, port := range ports {
+			result = append(result, corev1.ContainerPort{
+				ContainerPort: &port,
+			})
+		}
+		return result
+	}).(corev1.ContainerPortArrayOutput)
+
+	serviceDeployment.Deployment, err = appsv1.NewDeployment(ctx, name, &appsv1.DeploymentArgs{
+		Metadata: &metav1.ObjectMetaArgs{
+			Labels: args.Labels,
+			Name:   pulumi.String(name),
+		},
+		Spec: appsv1.DeploymentSpecArgs{
+			Selector: &metav1.LabelSelectorArgs{
+				MatchLabels: args.Labels,
+			},
+			Replicas: args.Replicas.ToIntOutput(),
+			Template: &corev1.PodTemplateSpecArgs{
+				Metadata: &metav1.ObjectMetaArgs{
+					Labels: args.Labels,
+				},
+				Spec: &corev1.PodSpecArgs{
+					Containers: corev1.ContainerArray{
+						corev1.ContainerArgs{
+							Name:      pulumi.String(name),
+							Image:     args.Image.ToStringOutput(),
+							Env:       args.Env.ToEnvVarArrayOutput(),
+							Resources: args.Resources.ToResourceRequirementsOutput(),
+							Ports:     deploymentPorts,
+						}},
+				},
+			},
+		},
+	}, pulumi.Parent(serviceDeployment))
+	if err != nil {
+		return nil, err
+	}
+
+	servicePorts := portsOutput.ApplyT(
+		func(ports []int) []corev1.ServicePort {
```

Comment:
Perhaps push this up onto previous line so that the rest of the nesting is simpler.